File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.cpp
// class Solution {
// public:
//     double findMaxAverage(vector<int>& nums, int k) {
//         int tmpSum = std::accumulate(nums.begin(), nums.begin()+k,0);
//         int maxSum = tmpSum;
//         for (int i = 0; i < nums.size()-k; i++){
//             tmpSum = tmpSum - nums[i] + nums[i+k];
//             maxSum = max(maxSum, tmpSum);
//         }
//         return (double)maxSum/k;
//     }
// };

class Solution {
public:
    double findMaxAverage(vector<int>& nums, int k) {
        if (nums.size() == 1) {
            return static_cast<double>(nums[0]);
        }
        
        int start = 0;
        int end = k;
        double average = 0.0;
        
        // Calculate the average of the first window
        for (int i = 0; i < k; i++) {
            average += static_cast<double>(nums[i]) / k;
        }
        
        // Initialize the maximum average
        double max_average = average;
        
        // Slide the window and update the maximum average
        while (end < nums.size()) {
            average = average - static_cast<double>(nums[start]) / k;
            average = average + static_cast<double>(nums[end]) / k;
            max_average = max(max_average, average);
            start++;
            end++;
        }
        
        return max_average;
    }
};

File: 0643-maximum-average-subarray-i/0643-maximum-average-subarray-i.py
class Solution:
    def findMaxAverage(self, nums: List[int], k: int) -> float:
        # added after looking at solution by deepankyadav
        if len(nums)==1:
            return nums[0]
        
        # my solution
        tmpSum = sum(nums[:k])
        maxSum = tmpSum
        for i in range(0, len(nums)-k):
            tmpSum = tmpSum-nums[i]+nums[i+k]
            maxSum = max(maxSum, tmpSum)
        return maxSum/k